httprequesthandler_test.go

```diff
@@ -0,0 +1,192 @@
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if r.Body != nil {
+				defer r.Body.Close()
+			}
+			if bodyBytes, err := ioutil.ReadAll(r.Body); err == nil {
+				actualBody = bodyBytes
+			}
+			actualContentLength = r.ContentLength
+		}))
+
+	wc := config.WatchdogConfig{
+		FunctionProcess: "sleep 10",
+		ExecTimeout:     time.Duration(time.Second * 10),
+		UpstreamURL:     upstreamServer.URL,
+	}
+
+	recordHTTPRequest(t, wc, http.MethodGet, bytes.NewBuffer(expectedBody), nil)
+
+	if !bytes.Equal(actualBody, expectedBody) {
+		t.Errorf("upstream received incorrect body - want: %s, got: %s", expectedBody, actualBody)
+	}
+
+	if actualContentLength != expectedContentLength {
+		t.Errorf("upstream received incorrect Content-Length - want: %d, got %d", expectedContentLength, actualContentLength)
+	}
+}
+
+func TestHTTPRequestHandler_WatchdogPassesHeaderToUpsteam(t *testing.T) {
+	expectedContentType := "text/plain"
+	actualContentType := ""
+	upstreamServer := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			actualContentType = r.Header.Get("Content-Type")
+		}))
+
+	wc := config.WatchdogConfig{
+		FunctionProcess: "sleep 10",
+		ExecTimeout:     time.Duration(time.Second * 10),
+		UpstreamURL:     upstreamServer.URL,
+	}
+
+	recordHTTPRequest(t, wc, http.MethodGet, nil, map[string]string{
+		"Content-Type": expectedContentType,
+	})
+
+	if actualContentType != expectedContentType {
+		t.Errorf("upstream received incorrect content-type - want: %s, got: %s", expectedContentType, actualContentType)
+	}
+}
+
+func TestHTTPRequestHandler_WatchdogReceivesStatusFromUpstream(t *testing.T) {
+	expectedStatus := http.StatusOK
+	upstreamServer := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(expectedStatus)
+		}))
+
+	wc := config.WatchdogConfig{
+		FunctionProcess: "sleep 10",
+		ExecTimeout:     time.Duration(time.Second * 10),
+		UpstreamURL:     upstreamServer.URL,
+	}
+
+	rr := recordHTTPRequest(t, wc, http.MethodGet, nil, nil)
+
+	if status := rr.Code; status != expectedStatus {
+		t.Errorf("handler returned wrong status code -  want: %v, got: %v", expectedStatus, status)
+	}
+}
+
+func TestHTTPRequestHandler_WatchdogReceivesBodyFromUpstream(t *testing.T) {
+	expectedBody := []byte("openfaas rocks")
+	upstreamServer := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Write(expectedBody)
+			w.WriteHeader(http.StatusOK)
+		}))
+
+	wc := config.WatchdogConfig{
+		FunctionProcess: "sleep 10",
+		ExecTimeout:     time.Duration(time.Second * 10),
+		UpstreamURL:     upstreamServer.URL,
+	}
+
+	rr := recordHTTPRequest(t, wc, http.MethodGet, nil, nil)
+
+	if rr.Body == nil {
+		t.Errorf("handler retured nil body: expected %s", string(expectedBody))
+	}
+
+	bodyBytes, bodyErr := ioutil.ReadAll(rr.Body)
+	if bodyErr != nil {
+		t.Error("unable to read data from body")
```

Comment:
perhaps if there is a "fatal" error also break / return

Reply:
When calling `t.Fatal`, it should stop the test and not continue running?